File: irrgarten-ruby/game.rb
# encoding: utf-8
require_relative 'dice'
require_relative 'player'
require_relative 'labyrinth'
require_relative 'monster'
require_relative 'directions'
require_relative 'orientation'
require_relative 'game_state'

module Irrgarten
  class Game
    @@MAX_ROUNDS = 10

    def initialize(n_players)
      @players = []
      @monsters = []
      @log = ""

      (0...n_players).each do |i|
        @players << Player.new(i.to_s, Dice.random_intelligence, Dice.random_strength) # << = @players.push = @players.append
      end

      @current_player_index = Dice.who_starts(n_players)
      @current_player = @players[@current_player_index]
      @log = ""

      @labyrinth = Labyrinth.new(10, 10, 0, 0)
      configure_labyrinth
      @labyrinth.spread_players(@players)
    end

    def finished
      @labyrinth.have_a_winner
    end

    def next_step(preferred_direction)
      if !@current_player.dead
        direction = actual_direction(preferred_direction)
        if direction != preferred_direction
          log_player_no_orders
        end

        puts direction
        monster = @labyrinth.put_player(direction, @current_player)

        if monster.nil?
          log_no_monster
        else
          winner = combat(monster)
          manage_reward(winner)
        end
      else
        manage_resurrection
      end

      end_game = finished

      unless end_game
        next_player
      end

      end_game
    end

    def game_state
      GameState.new(
        @labyrinth.to_s,
        @players.to_s,
        @monsters.to_s,
        @current_player_index,
        finished,
        @log
      )
    end

    private

    def configure_labyrinth
      m1 = Monster.new("bicho mostro", Dice.random_intelligence, Dice.random_strength)
      @monsters.append(m1)
      @labyrinth.add_monster(1, 0, m1)
      m2 = Monster.new("mostro bicho", Dice.random_intelligence, Dice.random_strength)
      @monsters.append(m2)
      @labyrinth.add_monster(2, 1, m2)
    end

    def next_player
      @current_player_index = (@current_player_index + 1) % @players.size
      @current_player = @players[@current_player_index]
    end

    def actual_direction(preferred_direction)
      current_row = @current_player.row
      current_col = @current_player.col
      valid_moves = @labyrinth.valid_moves(current_row, current_col)

      @current_player.move(preferred_direction, valid_moves)
    end

    def combat(monster)
      rounds = 0
      winner = GameCharacter::PLAYER
      player_attack = @current_player.attack
      lose = monster.defend(player_attack)

      while !lose && rounds < @@MAX_ROUNDS
        winner = GameCharacter::MONSTER
        rounds += 1
        monster_attack = monster.attack
        lose = @current_player.defend(monster_attack)

        if !lose
          player_attack = @current_player.attack
          winner = GameCharacter::PLAYER
          lose = monster.defend(player_attack)
        end
      end

      self.log_rounds(rounds, @@MAX_ROUNDS)
      winner
    end

    def manage_reward(winner)
      if winner == GameCharacter::PLAYER
        @current_player.receive_reward
        log_player_won
      else
        log_monster_won
      end
    end

    def manage_resurrection
      if Dice.resurrect_player
        @current_player.resurrect
        log_resurrected
      else
        log_player_skip_turn
      end
    end

    def log_player_won
      @log += "Player #{@current_player} won\n"
    end

    def log_monster_won
      @log += "Monster won the combat\n"
    end

    def log_resurrected
      @log += "Player #{@current_player} resurrected\n"
    end

    def log_player_skip_turn
      @log += "Player #{@current_player} skip turn\n"
    end

    def log_player_no_orders
      @log += "Player #{@current_player} didn't follow human player instructions\n"
    end

    def log_no_monster
      @log += "#{@current_player} empty cell or unable to move\n"
    end

    def log_rounds(rounds, max)
      @log += "Rounds: #{rounds} of #{max}\n"
    end
  end
end

File: irrgarten-java/src/irrgarten/Game.java
package irrgarten;

import java.util.ArrayList;
import java.util.List;

public class Game {

    private static final int MAX_ROUNDS = 10;
    private int currentPlayerIndex;
    private String log;

    private Player currentPlayer;
    private List<Player> players;
    private Labyrinth labyrinth;
    private List<Monster> monsters;

    public Game(int nPlayers) {
        // 0 < nPlayers < 10
        monsters = new ArrayList();
        players = new ArrayList(nPlayers);
        for (int i = 0; i <= nPlayers; i++) {
            players.add(new Player((char) (i + '0'), Dice.randomIntelligence(), Dice.randomStrength()));
        }
        currentPlayerIndex = Dice.whoStarts(nPlayers);
        currentPlayer = players.get(currentPlayerIndex);
        log = "";

        labyrinth = new Labyrinth(10, 10, 1, 1);
        configureLabyrinth();
        labyrinth.spreadPlayers(players.toArray(new Player[players.size()]));
    }

    public boolean finished() {
        return labyrinth.haveAWinner();
    }

    public boolean nextStep(Directions preferredDirection) {
        if (!currentPlayer.dead()) {
            Directions direction = actualDirection(preferredDirection);
            if (direction != preferredDirection) {
                logPlayerNoOrders();
            }
            Monster monster = labyrinth.putPlayer(direction, currentPlayer);
            
            if (monster == null) {
                logNoMonster();
            } else {
                GameCharacter winner = combat(monster);
                manageReward(winner);
            }
        
        } else {
            manageResurrection();
        }
        
        boolean endGame = finished();
        if (!endGame) {
            nextPlayer();
        }
        return endGame;
    }
    
    public GameState getGameState() {
        return new GameState(
                labyrinth.toString(),
                players.toString(),
                monsters.toString(),
                currentPlayerIndex,
                finished(),
                log
        );
    }

    private void configureLabyrinth() {
        // TODO: esto se puede hacer con posiciones aleatorias
        Monster m1 = new Monster("1", Dice.randomIntelligence(), Dice.randomStrength());
        monsters.add(m1);
        labyrinth.addMonster(1, 0, m1);
        m1 = new Monster("1", Dice.randomIntelligence(), Dice.randomStrength());
        monsters.add(m1);
        labyrinth.addMonster(2, 1, m1);
        
        labyrinth.addBlock(Orientation.HORIZONTAL, 0, 0, 10);
    }

    private void nextPlayer() {
        currentPlayerIndex = (currentPlayerIndex + 1) % players.size();
        currentPlayer = players.get(currentPlayerIndex);
    }
    
    private Directions actualDirection(Directions preferredDirection) {
        int currentRow = currentPlayer.getRow();
        int currentCol = currentPlayer.getCol();
        Directions[] validMoves = labyrinth.validMoves(currentRow, currentCol);
        return currentPlayer.move(preferredDirection, validMoves);
    }

    private GameCharacter combat(Monster monster) {
        int rounds = 0;
        GameCharacter winner = GameCharacter.PLAYER;
        
        float playerAttack = currentPlayer.attack();
        boolean lose = monster.defend(playerAttack);
        while (!lose && rounds < MAX_ROUNDS) {
            winner = GameCharacter.MONSTER;
            rounds++;
            float monsterAttack = monster.attack();
            lose = currentPlayer.defend(monsterAttack);
            
            if (!lose) {
                playerAttack = currentPlayer.attack();
                winner = GameCharacter.PLAYER;
                lose = monster.defend(playerAttack);
            }
        }
        
        logRounds(rounds, MAX_ROUNDS);

        return winner;
    }
    
    private void manageReward(GameCharacter winner) {
        if (winner == GameCharacter.PLAYER) {
            currentPlayer.receiveReward();
            logPlayerWon();
        } else {
            logMonsterWon();
        }        
    }
    
    private void manageResurrection() {
        if (Dice.resurrectPlayer()) {
            currentPlayer.resurrect();
            logResurrected();
        } else {
            logPlayerSkipTurn();
        }
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    private void logPlayerWon() {
        log += "Player " + currentPlayer.getNumber() + " wins\n";
    }

    private void logMonsterWon() {
        log += "Monster wins the combat\n";
    }

    private void logResurrected() {
        log += "Player " + currentPlayer.getNumber() + " resurrects\n";
    }

    private void logPlayerSkipTurn() {
        log += "Player " + currentPlayer.getNumber() + " skips their turn\n";
    }

    private void logPlayerNoOrders() {
        log += "Player " + currentPlayer.getNumber() + " doesn't follow human player instructions\n";
    }

    private void logNoMonster() {
        log += "Player " + currentPlayer.getNumber() + " moves to an empty cell or is unable to move\n";
    }

    private void logRounds(int rounds, int max) {
        log += rounds + "/" + max + " combat rounds played\n";
    }

}
